runtime/connectors/blob/blobdownloader.go

```diff
@@ -0,0 +1,121 @@
+package blob
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"github.com/bmatcuk/doublestar/v4"
+	"github.com/rilldata/rill/runtime/pkg/fileutil"
+
+	"gocloud.dev/blob"
+)
+
+type FetchConfigs struct {
+	MaxSize       int64
+	MaxDownload   int
+	MaxIterations int64
+}
+
+func blobType(path string) BlobType {
+	if strings.Contains(path, "file") {
+		return File
+	} else if strings.Contains(path, "gs") {
+		return GCS
+	} else if strings.Contains(path, "s3") {
+		return S3
+	}
+	return File
+}
+
+func FetchBlobHandler(ctx context.Context, bucket *blob.Bucket, config FetchConfigs, globPattern, bucketPath string) (*BlobHandler, error) {
+	prefix, glob := doublestar.SplitPattern(globPattern)
+	result := &BlobHandler{prefix: prefix, bucket: bucket, BlobType: blobType(bucketPath), path: bucketPath}
+	if !fileutil.HasMeta(glob) {
+		// glob represent plain object
+		result.FileNames = []string{globPattern}
+		return result, nil
+	}
+	before := func(as func(interface{}) bool) error {
+		// Access storage.Query via q here.
+		var q *storage.Query
+		if as(&q) {
+			// we only need name and size, adding only required attributes to reduce data fetched
+			_ = q.SetAttrSelection([]string{"Name", "Size"})
+		}
+		return nil
+	}
+
+	listOptions := blob.ListOptions{BeforeList: before}
+	if prefix != "." {
+		listOptions.Prefix = prefix
+	}
+
+	var size int64
+	matchCount := 0
+	fileNames := make([]string, 0)
+	// list max matched files or 100 in one API listing
+	pageSize := int(math.Max(100, float64(config.MaxDownload)))
+	fetched := int64(0)
+	var returnErr error
+	validateConfigs(&config)
+	for token := blob.FirstPageToken; token != nil; {
+		iter, nextToken, err := bucket.ListPage(ctx, token, pageSize, &listOptions)
+		if err != nil {
+			returnErr = err
+			break
+		}
+		token = nextToken
+		for _, obj := range iter {
+			if match(glob, obj.Key) {
+				size += obj.Size
+				matchCount++
+				fileNames = append(fileNames, obj.Key)
+			}
+		}
+
+		if err := validateLimits(size, matchCount, fetched, config); err != nil {
+			returnErr = err
+			break
+		}
+	}
+	if returnErr != nil {
+		bucket.Close()
+		return nil, returnErr
+	}
```

Comment:
In normal scenario we expect client to close the bucket once files are downloaded, we are closing buckets only in error scenarios here.

Reply:
Even in error scenarios, the responsibility of closing should be on the opener (unless explicitly documented, e.g. if the opener hands over that responsibility to somewhere else).

A `defer` statement will also cover error cases. E.g. imagine this code:
```go
// Open something
x, err := thing.Open()
if err != nil {
  return err
}
// Since it opened successfully, we're now responsible for closing it
defer x.Close()

// Use x
err = x.DoSomething()
if err != nil {
  return err // x.Close() will be triggered here due to defer
}

// Success
return nil // x.Close() will be triggered here due to defer
```